### PR TITLE
[Feature] feat: add video frame interpolation postprocess

### PR DESCRIPTION
**Pull request by david6666666**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

- ## Purpose
Add post-generation frame interpolation for the `/v1/videos` and `/v1/videos/sync` video generation APIs, aligned with sglang's RIFE 4.22.lite postprocess behavior.

This adds request parameters for enabling RIFE interpolation, applies interpolation before MP4 encoding, scales output FPS by `2**frame_interpolation_exp`, and documents the video API usage. The diffusion denoise/scheduler path is unchanged.

- ## Test Plan

Serve command:

```bash
vllm serve Wan-AI/Wan2.2-T2V-A14B-Diffusers --omni --port 8091 --enable-diffusion-pipeline-profiler
```
Curl command:

```
curl -X POST http://localhost:8091/v1/videos/sync \
  -F "prompt=A dog running through a park" \
  -F "num_frames=81" \
  -F "width=832" \
  -F "height=480" \
  -F "fps=16" \
  -F "num_inference_steps=40" \
  -F "guidance_scale=4.0" \
  -F "guidance_scale_2=4.0" \
  -F "enable_frame_interpolation=true" \
  -F "frame_interpolation_exp=1" \
  -F "frame_interpolation_scale=1.0" \
  -F "seed=42" \
  -o sync_t2v_interpolated.mp4
```

- ## Test Result
AMD MI300X

```
INFO 04-09 11:31:03 [diffusion_pipeline_profiler.py:31] [DiffusionPipelineProfiler] Wan22Pipeline.vae.decode took 5.327326s
INFO 04-09 11:31:03 [diffusion_pipeline_profiler.py:31] [DiffusionPipelineProfiler] Wan22Pipeline.forward took 254.884285s
INFO 04-09 11:31:03 [diffusion_model_runner.py:212] Peak GPU memory (this request): 76.63 GB reserved, 73.47 GB allocated, 3.16 GB pool overhead (4.1%)
INFO 04-09 11:31:03 [diffusion_engine.py:126] Generation completed successfully.
INFO 04-09 11:31:03 [weight_utils.py:50] Using model weights format ['flownet.pkl']
INFO 04-09 11:31:04 [rife_interpolator.py:276] Loaded RIFE weights from /root/.cache/huggingface/hub/models--elfgum--RIFE-4.22.lite/snapshots/99d6892a9f4c039cb37ff21c9530e79b13f0b30b/flownet.pkl
INFO 04-09 11:31:04 [rife_interpolator.py:379] RIFE model loaded on device: cuda
INFO 04-09 11:31:09 [diffusion_engine.py:169] Post-processing completed in 5.2878 seconds
INFO 04-09 11:31:09 [diffusion_engine.py:172] DiffusionEngine.step breakdown: preprocess=0.01 ms, add_req_and_wait=255318.23 ms, postprocess=5287.81 ms, total=260606.43 ms
(APIServer pid=21842) INFO 04-09 11:31:11 [omni_base.py:162] [Summary] {}
(APIServer pid=21842) INFO 04-09 11:31:12 [serving_video.py:213] Video response encoding (MP4 bytes): 1116.16 ms
```

```
root@rocm-7-2-software-gpu-mi300x1-192gb-devcloud-atl1:/workspace/vllm-omni# ffprobe -v error -select_streams v:0 -show_entries stream=r_frame_rate,nb_frames,duration -of default=noprint_wrappers=1 sync_t2v.mp4 
r_frame_rate=16/1
duration=5.062500
nb_frames=81
root@rocm-7-2-software-gpu-mi300x1-192gb-devcloud-atl1:/workspace/vllm-omni# ffprobe -v error -select_streams v:0 -show_entries stream=r_frame_rate,nb_frames,duration -of default=noprint_wrappers=1 sync_t2v_interpolated.mp4 
r_frame_rate=32/1
duration=5.031250
nb_frames=161

```

Before frame interpolation

https://github.com/user-attachments/assets/12490222-cc64-4567-ad2b-eb6c4e83dd85

After frame interpolation

https://github.com/user-attachments/assets/90018f57-e61e-46b8-bdad-b3c85b5bf648

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by david6666666** [comment]
@SamitHuang @wtomin PTAL thx

File: vllm_omni/diffusion/postprocess/rife_interpolator.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+
+        with _MODEL_CACHE_LOCK:
+            if resolved_path in _MODEL_CACHE:
+                return _MODEL_CACHE[resolved_path]
+
+            device = _select_torch_device()
+            model = Model()
+            model.load_model(resolved_path)
+            model.eval()
+            model.flownet = model.flownet.to(device)
```
Comment (SamitHuang) [inline vllm_omni/diffusion/postprocess/rife_interpolator.py:377]: Loading a PyTorch model onto the GPU within the API server process initializes a CUDA context here. In vLLM-Omni, the API server is designed to be lightweight. The inference workers already allocate most of the GPU memory, so this will likely lead to OOM errors.

File: vllm_omni/entrypoints/openai/serving_video.py [hunk 2]
```diff
@@ -165,17 +165,19 @@ async def generate_videos(
         videos, audios, audio_sample_rate, output_fps = await self._run_and_extract(
             request, reference_id, reference_image=reference_image
         )
+        frame_interpolation_kwargs = self._frame_interpolation_kwargs(request)
         _t_encode_start = time.perf_counter()
         video_data = [
             VideoData(
                 b64_json=(
-                    encode_video_base64(video, fps=output_fps)
+                    encode_video_base64(video, fps=output_fps, **frame_interpolation_kwargs)
```
Comment (SamitHuang) [inline vllm_omni/entrypoints/openai/serving_video.py]: This synchronous call triggers PyTorch inference. Because it runs directly in the async handler, it will block the entire `asyncio` event loop, hanging all other API requests. Wrap this in an async executor (e.g., `asyncio.to_thread`).

**Review by SamitHuang — CHANGES_REQUESTED** [review]
This feature introduces a major architectural flaw: running heavy PyTorch inference (RIFE) directly in the API server process.

1. **OOM Risk:** This initializes a CUDA context in the main API process, which will unexpectedly consume VRAM and likely cause OOM issues when `gpu_memory_utilization` is high.
2. **Event Loop Blocking:** The inference is fully synchronous. Running it here will block the FastAPI `asyncio` event loop, causing all other incoming requests to hang.

Frame interpolation must be moved to an asynchronous executor or delegated to the worker processes. Please provide memory measurements showing the peak VRAM impact when this feature is enabled. Additionally, the current tests only mock `interpolate_video_frames`—please add a unit test that runs the actual PyTorch interpolation logic on dummy tensors.

File: vllm_omni/diffusion/postprocess/rife_interpolator.py (hunk 1 above)
Comment (david6666666) [reply vllm_omni/diffusion/postprocess/rife_interpolator.py:377]: Good catch. I changed the design so frame interpolation no longer runs in the API server process. It is now executed in the Wan worker-side postprocess path, so it reuses the diffusion worker's device/context instead of creating a separate GPU context in the API server. The API server now only receives the already-interpolated video plus an FPS multiplier and performs MP4 export.

File: vllm_omni/entrypoints/openai/serving_video.py (hunk 2 above)
Comment (david6666666) [reply vllm_omni/entrypoints/openai/serving_video.py]: Agreed. I kept the API-server-side encoding path off the event loop by wrapping the MP4/base64 export calls in `asyncio.to_thread(...)`. After the refactor, the heavy interpolation logic also moved out of the async handler and into the diffusion worker postprocess path.

**Comment by david6666666** [comment]
Addressed the main review concerns in the latest updates:

## What changed

- Moved frame interpolation out of the API server process and into the Wan worker-side postprocess path.
- Kept API-server-side video encoding off the FastAPI event loop via `asyncio.to_thread(...)`.
- Propagated interpolation controls through `OmniDiffusionSamplingParams` so the worker postprocess can make the interpolation decision.
- Returned the FPS multiplier from worker postprocess through `custom_output`, so the API server only applies the final output FPS during MP4 export.
- Added actual PyTorch unit coverage for the RIFE path on dummy tensors (`Model.inference(...)` and `FrameInterpolator.interpolate_tensor(...)`), instead of only mocking the interpolation call.

## Note on memory measurement

I have not attached a dedicated peak-VRAM measurement in this reply yet. The architectural change here is specifically to avoid introducing a second GPU context in the API server process; interpolation now runs in the existing worker-side execution path. If needed, I can follow up with a focused memory measurement from a runtime environment.

File: vllm_omni/diffusion/postprocess/rife_interpolator.py [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+    """Lazy-loaded RIFE 4.22.lite frame interpolator."""
+
+    def __init__(self, model_path: str | None = None):
+        self._model_path = model_path
+        self._resolved_path: str | None = None
+
+    def _ensure_model_loaded(self) -> Model:
+        resolved_path = _resolve_rife_model_path(self._model_path)
+        self._resolved_path = resolved_path
+        device = _select_torch_device()
```
Comment (david6666666) [inline vllm_omni/diffusion/postprocess/rife_interpolator.py]: `DiffusionEngine.step()` explicitly moves diffusion outputs to CPU before postprocess when `enable_cpu_offload` is on, because there may be no device headroom left for postprocess intermediates. This helper ignores that contract and always picks the accelerator again via `_select_torch_device()`, then moves the whole video back with `video.to(model.device())`. As a result, enabling frame interpolation can reintroduce the exact OOM that CPU offload was avoiding. Could we either keep interpolation on the tensor's current device, or explicitly force a CPU fallback when the diffusion output was offloaded? A regression test that covers `enable_cpu_offload + enable_frame_interpolation` would also help here.

**Review by david6666666 — COMMENTED** [review]
One blocking issue in the interpolation path.

**Review by SamitHuang — APPROVED** [review]
LGTM
